### ISSUE TITLE
Avatar images do not reliably fall back on a default when offline

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#8291]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Put your phone on airplane mode and turn off wifi
2. Close the app if it's already open.
3. Open or reopen the app

## Expected Result:
We should fallback on a default avatar for those images that are not cached

## Actual Result:
Avatar images do not reliably fall back on a default when offline

## Context: 
Related: the default avatar for a user is injected into their PersonalDetails [here](https://github.com/Expensify/Web-Expensify/blob/4fb872bb4bd85bc999e868b75ee5ef16c7b2bf18/lib/PersonalDetail.php#L95). Because it's just a URL, that default avatar image will only be available online (unless it happens to be cached in Onyx).
To fix that, we could create a DefaultAvatar svg icon and just pass a random fill to that component rather than using a URL for those icons. Might also have a minor performance improvement, especially on low-bandwidth connections on mobile.

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- iOS
- Android

**Version Number:** 1.1.43-2
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 
![IMG_4542](https://user-images.githubusercontent.com/43995119/159850404-078451ce-6460-4a6a-bf8c-85bdc40eb8f2.jpg)

**Expensify/Expensify Issue URL:** 
**Issue reported by:** 
**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by kbecciv** [comment]
@chiragsalian Also having the same issue when searched to invite a new member for Workspace. 

https://user-images.githubusercontent.com/93399543/159957813-b4305440-1a0f-4a1b-b28e-f83a10af2e0f.mp4

**Comment by chiragsalian** [comment]
Should be a pretty straightforward external issue.

**Comment by dylanexpensify** [comment]
Internal: https://www.upwork.com/ab/applicants/1508351478606524416/job-details
External: https://www.upwork.com/jobs/~01e341ebb7d774aad0

**Comment by Puneet-here** [comment]
### Proposal 
We can add a background color here -
https://github.com/Expensify/App/blob/50496a0a0d7dd283f9936988cda95abf69741342/src/components/Avatar.js#L42-L45

like this -

```diff
const imageStyle = [ 
     StyleUtils.getAvatarStyle(this.props.size), 
+     {backgroundColor: themeColors.icon},
     ...this.props.imageStyles, 
 ]; 
```
I found that we were doing this earlier but it was removed also I found that we were only using the background color when it isn't a chat room because when there's no chatroom icon the backround shows up as a gray square, but I don't know if there's a case where chatroom icon won't show up. So should we just add a background color or should be check if it's not a chatroom and only then we will add a background

**Comment by Santhosh-Sellavel** [comment]
> Related: the default avatar for a user is injected into their PersonalDetails [here](https://github.com/Expensify/Web-Expensify/blob/4fb872bb4bd85bc999e868b75ee5ef16c7b2bf18/lib/PersonalDetail.php#L95). Because it's just a URL, that default avatar image will only be available online (unless it happens to be cached in Onyx).
To fix that, we could create a DefaultAvatar svg icon and just pass a random fill to that component rather than using a URL for those icons. Might also have a minor performance improvement, especially on low-bandwidth connections on mobile.

@Puneet-here This is what we are looking for here.

**Comment by AndrewGable** [comment]
@michelle-thompson - Can we get the default avatar SVG for this? Thanks!

**Comment by dylanexpensify** [comment]
Just checking in, should this be doubled or are we feeling ok with where it's at? cc @Santhosh-Sellavel

**Comment by chiragsalian** [comment]
unassigned myself since just 1 CME should be enough.

**Comment by Santhosh-Sellavel** [comment]
Thanks for checking, @dylanexpensify. Let's wait on this one. 

> @michelle-thompson - Can we get the default avatar SVG for this? Thanks!

Any update on this?

**Comment by shawnborton** [comment]
Maybe we should just use the same style of default avatars we use, but use dark version or gray version as the fallback?
<img width="464" alt="image" src="https://user-images.githubusercontent.com/2319350/162246914-4501290a-1e3f-4c54-8bb0-1518f601c465.png">

Technically speaking though, how do we make sure that when an avatar is loading over a connection that the default version doesn't flash quickly before the real asset loads?

**Comment by shawnborton** [comment]
Sounds good to me. I think I learn towards the gray version given that this would only be shown as a fallback for offline... curious what the other designers think though.

**Comment by megankelso** [comment]
I really like the navy for icons but I agree that the grey might make more sense here for offline

**Comment by shawnborton** [comment]
Cool, let's do gray then. Here is the .svg we'll need: [fallback-avatar.svg.zip](https://github.com/Expensify/App/files/8446219/fallback-avatar.svg.zip)


Can we also make sure to use a background color for these avatar circles too? I'm not sure when they were removed, but we should have a default gray color there as well.

**Comment by dylanexpensify** [comment]
Another week passed, should we double? cc @Santhosh-Sellavel

**Comment by Santhosh-Sellavel** [comment]
Lets double.

**Comment by dylanexpensify** [comment]
Doubled!

**Comment by dylanexpensify** [comment]
Price doubled

**Comment by jayeshmangwani** [comment]
Proposal:

in [avatar](https://github.com/Expensify/App/blob/main/src/components/Avatar.js) file we can add [onError](https://reactnative.dev/docs/image#onerror) in Image component here

https://github.com/Expensify/App/blob/db6fab75722e833ca6e54f634e8a6b6f3c1cae79/src/components/Avatar.js#L53

when Image gets an error in load we can setState a error flag true and if its true then we have to  load the local SVG file, 
I have done this changes in local and its working , please check video here


https://user-images.githubusercontent.com/35371050/165837995-9c947c87-d818-4111-a4a1-727b32fab20e.mov

Updated [avatar](https://github.com/Expensify/App/blob/main/src/components/Avatar.js) file

```import React, {PureComponent} from 'react';
import {Image, View} from 'react-native';
import PropTypes from 'prop-types';
import _ from 'underscore';
import stylePropTypes from '../styles/stylePropTypes';
import Icon from './Icon';
import themeColors from '../styles/themes/default';
import CONST from '../CONST';
import * as StyleUtils from '../styles/StyleUtils';
import * as Expensicons from './Icon/Expensicons';
import defaultTheme from '../styles/themes/default';

const propTypes = {
    /** Source for the avatar. Can be a URL or an icon. */
    source: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),

    /** Extra styles to pass to Image */
    imageStyles: PropTypes.arrayOf(PropTypes.object),

    /** Extra styles to pass to View wrapper */
    containerStyles: stylePropTypes,

    /** Set the size of Avatar */
    size: PropTypes.oneOf(_.values(CONST.AVATAR_SIZE)),

    /** The fill color for the icon. Can be hex, rgb, rgba, or valid react-native named color such as 'red' or 'blue' */
    fill: PropTypes.string,
};

const defaultProps = {
    source: null,
    imageStyles: [],
    containerStyles: [],
    size: CONST.AVATAR_SIZE.DEFAULT,
    fill: themeColors.icon,
};

class Avatar extends PureComponent {
    constructor() {
        super();
        this.state = {
            isErrorFetchingImage: false, 
        }
        this.getUserImage = this.getUserImage.bind(this);
    }
    
    getUserImage() {
        const imageStyle = [
            StyleUtils.getAvatarStyle(this.props.size),
            ...this.props.imageStyles,
        ];

        if(this.state.isErrorFetchingImage) {
            let { height, width } = imageStyle;
            if(!height) {
                const {height: convertedHeight} = StyleUtils.getAvatarStyle(this.props.size)
                height = convertedHeight;
            };
            if(!width) {
                const {width: convertedWidth} = StyleUtils.getAvatarStyle(this.props.size)
                width = convertedWidth;
            }
            return (
                <Icon
                    src={Expensicons.FallbackAvatar}
                    height={height}
                    width={width}
                    fill={defaultTheme.offline}
                />
            )
        } 
        return (
            <Image
                onError={(err)=> this.setState({ isErrorFetchingImage:true })}
                source={{uri: this.props.source}}
                style={imageStyle}
                onLoad={()=> { if(this.state.isErrorFetchingImage){this.setState({ isErrorFetchingImage: false })}}}
            />
        )
    };

    render() {
        if (!this.props.source) {
            return null;
        }

        const iconSize = StyleUtils.getAvatarSize(this.props.size);
        
        return (
            <View pointerEvents="none" style={this.props.containerStyles}>
                {
                _.isFunction(this.props.source)
                    ? <Icon src={this.props.source} fill={this.props.fill} height={iconSize} width={iconSize} />
                    : this.getUserImage()
            }
            </View>
        );
    }
}

Avatar.defaultProps = defaultProps;
Avatar.propTypes = propTypes;
export default Avatar;
```

**Comment by Santhosh-Sellavel** [comment]
@jayeshmangwani seems your proposal would address this issue, but it's hard to follow.

Check out the closed [issues](https://github.com/Expensify/App/issues?q=is%3Aissue+is%3Aclosed+label%3A%22Awaiting+Payment%22) to how to write a proper proposal.

Can you resubmit your proposal, with the changes you are proposing in step by step manner? So it would be easy to understand. Make your code is DRY.
Ex: 
1. Should change the following lines in the Header.js file to fix this.
2. Add the following code here to get this done.

Thanks!

**Comment by jayeshmangwani** [comment]
> @jayeshmangwani seems your proposal would address this issue, but it's hard to follow.
> 
> Check out the closed [issues](https://github.com/Expensify/App/issues?q=is%3Aissue+is%3Aclosed+label%3A%22Awaiting+Payment%22) to how to write a proper proposal.
> 
> Can you resubmit your proposal, with the changes you are proposing in step by step manner? So it would be easy to understand. Make your code is DRY. Ex:
> 
> 1. Should change the following lines in the Header.js file to fix this.
> 2. Add the following code here to get this done.
> 
> Thanks!

@Santhosh-Sellavel  Thanks for checking the proposal, let me update my proposal in detail,

Proposal - updated

1. we have to create a state for checking errors in image loading, isErrorFetchingImage value that defines the error in loading the image, initially value will be false

```
constructor() {
  super();
    this.state = {
      isErrorFetchingImage: false, 
    }
  }
```
we have to add the above code here, 

https://github.com/Expensify/App/blob/f174c3223a1d6d0f571e5bd799441690951f078c/src/components/Avatar.js#L36-L37

2. [onError](https://reactnative.dev/docs/image#onerror,) method of Image component  that calls if the getting errors in loading image

```
onError={()=> this.setState({ isErrorFetchingImage: true })}
```
we have to add the above code here, 

https://github.com/Expensify/App/blob/f174c3223a1d6d0f571e5bd799441690951f078c/src/components/Avatar.js#L53


3. Icon is already being used in component, added a condition if the error flag will be true then also Icon will be visible, so either we getting function as a source or we are getting error in loading the remote URL Icon will be display



https://github.com/Expensify/App/blob/f174c3223a1d6d0f571e5bd799441690951f078c/src/components/Avatar.js#L51

above line will be replaced with this line of code

```
_.isFunction(this.props.source) || this.state.isErrorFetchingImage
```

4. For Icon we have to add 2 conditions for src and fill, for image error src will be fallback-avatar.svg and fill will be defaultTheme.offline


https://github.com/Expensify/App/blob/f174c3223a1d6d0f571e5bd799441690951f078c/src/components/Avatar.js#L52

above code will be replaced with this code

```
    
<Icon
    src={this.state.isErrorFetchingImage ? Expensicons.FallbackAvatar : this.props.source }
    fill={this.state.isErrorFetchingImage ? defaultTheme.offline : this.props.fill}
    height={iconSize}
    width={iconSize}
/>
```

Thanks!

**Comment by Santhosh-Sellavel** [comment]
@AndrewGable
@jayeshmangwani [proposal](https://github.com/Expensify/App/issues/8291#issuecomment-1114014660) LGTM!
🎀👀🎀 C+ reviewed

@jayeshmangwani 
I think this should be,
```js
constructor(props) {
  super(props);
```
Could have used `themeColors` which is already available instead of `defaultTheme`.

**Comment by jayeshmangwani** [comment]
@Santhosh-Sellavel  Thanks for the suggestion, now I have updated these 2 changes in the code

**Comment by jayeshmangwani** [comment]
PR raised @Santhosh-Sellavel 
CC: @AndrewGable

**Comment by Santhosh-Sellavel** [comment]
@bfitzexpensify Can you set up a job on Upwork for this issue or share me link for this one!

**Comment by bfitzexpensify** [comment]
Invited you to the current open job @Santhosh-Sellavel

**Comment by bfitzexpensify** [comment]
Payments issued - thanks everyone for the work here!
